hips/hip-0012.md

```diff
@@ -0,0 +1,178 @@
+
+Describe why particular design decisions were made.
+
+## Specification
+
+### The Release Engineer
+
+A single individual will oversee the planning and development phases of Helm 4.
+This person will be the _Release Engineer_.
+The person will have the following responsibilities:
+
+- Lead the kick-off meeting
+- Define the process for including features
+- Determine the timelines for development
+- Make the determination of when the project has indeed reached a milestone
+- Name and oversee releases
+- After the release, the release manager will determine what the "intent" of a Helm behavior was (which informs determining whether an incoming breaking change is a feature or a bugfix)
+
+#### Determining Who Will Be Release Engineer
+
+The release engineer must be a core maintainer on the `github.com/helm/helm` project.
+The engineer must not be in danger of losing core maintainership due to inactivity.
+Anyone can nominate a core maintainer to become the release engineer (self-nominations are encouraged).
+Any nominee must _agree_ to the nomination.
+The Helm project maintainers will determine which of the nominated maintainers shall be the release engineer.
+
+In the event that a release engineer cannot perform the duties required,
+the Helm project maintainers may choose a replacement.
+
+### The Process
+
+There will be three phases to Helm 4 development:
+
+1. Planning
+2. Feature development
+3. Release preparation
+
+The timeline will be approximately:
+
+- Naming of a Release Engineer: Sept. 2021
+- Kick-off meeting: Oct. 2021
+- Earliest that engineering can begin: Jan. 2022
+- Earliest that a release could possibly be: May, 2022
+
+More likely, the release will be late 2022 to midyear 2023.
+
+#### Planning
+
+This phase begins when the release engineer is named.
+The release engineer shall schedule a kick-off meeting in which all Helm maintainers (from any Helm org project) will be invited to participate.
+Beyond these participants, the release engineer may determine who else will be included, and may determine the meeting to be public.
+The kick-off meeting will open the floor to feature ideas and changes, with preference toward ideas already documented as HIPs.
+
+This process will continue as long as the release engineer determines necessary,
+but not less than three months.
+This minimum time frame is intended to give ample time for the community to provide input.
+
+All features and breaking changes MUST be described in a HIP, and that HIP must pass the usual process.
+This provides three benefits:
+
+- All changes are documented
+- All changes have had a period of scrutiny prior to the PR being written
+- There is a clear record of alternative considerations
+
+Things that are deemed by core maintainers to be "bug fixes" (and not changes or features) do not need HIPs.
+
+#### Feature Development
+
+During this phase, the features and changes agreed upon during planning will be implemented in a development branch of Helm.
```

Comment:
Just wondering if it is worth considering using the default/main branch instead. My reasoning is as follows:

- You get to pull up features and bug fixes from Helm 3 branch as you go along.Thereby staying in sync with Helm 3 branch and avoid "big bang" approach at end of Helm 4 cycle of trying to pull in all changes from Helm 3 since branch split.
- Helps minimize missed capability from Helm 3. This is a regular occurrence in Helm 3 with issues raised of missing capability from Helm 2.
- Avoids having to swap branches (helm4-dev to main/main to helm3) when Helm 4 is released. This caused some confusion for Helm 3.

Reply:
It's also worth noting that Go uses the default branch for versioning. We will likely need to be a little more careful about how we develop Helm 4 as many users rely on `helm.sh/helm` APIs for their projects.

The migration from Helm 2 to Helm 3 was relatively simple for us because Go had no notion of versioning at that time and we weren't concerned as much about version management. But that may likely change with Helm 4.

Reply:
Ok, should this be added to the doc then?

Reply:
I dunno... I think it may be worth to write this up as a separate HIP. HIP 12 is more about how Helm 4 is going to be planned - specific implementation details about said plan seems more appropriate as a separate HIP (see section on "Planning").

IOW I'm okay with moving ahead here, but we should probably do some research, then document how Helm expects to handle major version upgrades to the Helm SDK in another HIP (and notify the release engineer which maintainer is planning to be responsible for that work). Once we have a better idea of that plan, we can update HIP 12.

Reply:
@bacongobbler Ok, it might be worth having a different HIP for specific details. Overall, HIP LGTM and we can work out other details  next.